### PR TITLE
Plugins: exposing PanelQueryRunner from grafana/runtime so it can be used in plugins.

### PR DESCRIPTION
**Pull request by mckn**
**What this PR does / why we need it**:
In this PR I have moved the typings for PanelQueryRunner to `grafana/data` and I am exposing it via `grafana/runtime` so it can be reusing by plugins.

The main reason for doing this is that we want to use it in an external app.

**Special notes for your reviewer**:
This is the easies way of doing this that I could think of. I'm not sure we should expose all of the methods I have added to the QueryRunner interface so feel free to have an extra look at that.

*Question:* Should I rename `GetDataOptions` to `QueryRunnerGetDataOptions` since it is exposed via `grafana/data`?

### CODE REVIEW

**Comment by mckn** [comment]
I will add proper code comments after I have got feedback on this PR.

**Review by hugohaggmark — APPROVED** [review]
Looks great 🎉 , I think the naming part is hard. We're going to introduce a `DashboardQueryRunner` soon, but that might just reuse the `QueryRunner` interface not sure, but it may also introduce another interface and then `QueryRunner` seems less good. As always this is tricky.

File: packages/grafana-data/src/types/queryRunner.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  timeInfo?: string; // String description of time range for display
+  maxDataPoints: number;
+  minInterval: string | undefined | null;
+  scopedVars?: ScopedVars;
+  cacheTimeout?: string;
+  delayStateNotification?: number; // default 100ms.
+  transformations?: DataTransformerConfig[];
+}
+
+export interface GetDataOptions {
```
Comment (torkelo) [inline packages/grafana-data/src/types/queryRunner.ts:27]:
```suggestion
export interface QueryRunnerGetDataOptions {
```

**Comment by dprokop** [comment]
@mckn I see empty `packages/grafana-runtime/src/queryRunners/useQueryRunner.ts` file in this PR - something you messed committing or haven't implemented yet?

File: packages/grafana-data/src/types/queryRunner.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { Observable } from 'rxjs';
+import { DataQuery, DataSourceApi, DataSourceJsonData } from './datasource';
+import { PanelData } from './panel';
+import { ScopedVars } from './ScopedVars';
+import { TimeRange, TimeZone } from './time';
+import { DataTransformerConfig } from './transformations';
+
+export interface QueryRunnerOptions<
```
Comment (ryantxu) [inline packages/grafana-data/src/types/queryRunner.ts:8]: Do we need to add this to `@grafana/data`?  how about adding it just in the `@grafana/runtime` package.  In runtime I feel a bit more comfortable if we need to change/evolve the interface to match the current runtime system.  In data, we should *try* to keep it as stable as possible

Are the generics actually useful with this interface?  I think they all end as DataQuery/DataSourceJsonData anyway.

File: packages/grafana-data/src/types/queryRunner.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  delayStateNotification?: number; // default 100ms.
+  transformations?: DataTransformerConfig[];
+}
+
+export interface GetDataOptions {
+  withTransforms: boolean;
+  withFieldConfig: boolean;
+}
+
+export interface QueryRunner {
```
Comment (ryantxu) [inline packages/grafana-data/src/types/queryRunner.ts:32]: Do all these functions need to be exposed?  Can some of them stay internal?

File: packages/grafana-data/src/types/queryRunner.ts (hunk 1 above)
Comment (mckn) [reply packages/grafana-data/src/types/queryRunner.ts:27]: Super! I will change this 👍

File: packages/grafana-data/src/types/queryRunner.ts (hunk 2 above)
Comment (mckn) [reply packages/grafana-data/src/types/queryRunner.ts:8]: hmm, we might be able to change the interface but let's do that in a different PR. I would like to keep this as minimal as possible.

So the benefit of having this in `@grafana/data` is that we can have a dependency on this interface in e.g. `@grafana/ui` which might be beneficial. We had this problem with the `DataSourceSrv`. If that interface would have been in `@grafana/data` we could have moved that component to `@grafana/ui`.

**Comment by mckn** [comment]
> Looks great 🎉 , I think the naming part is hard. We're going to introduce a `DashboardQueryRunner` soon, but that might just reuse the `QueryRunner` interface not sure, but it may also introduce another interface and then `QueryRunner` seems less good. As always this is tricky.

hmm, could we name it something else..do you have any suggestions?

File: packages/grafana-data/src/types/queryRunner.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  dashboardId?: number;
+  timezone: TimeZone;
+  timeRange: TimeRange;
+  timeInfo?: string; // String description of time range for display
+  maxDataPoints: number;
+  minInterval: string | undefined | null;
+  scopedVars?: ScopedVars;
+  cacheTimeout?: string;
+  delayStateNotification?: number; // default 100ms.
+  transformations?: DataTransformerConfig[];
```
Comment (torkelo) [inline packages/grafana-data/src/types/queryRunner.ts:30]: You can remove this prop, it's not used

File: packages/grafana-data/src/types/queryRunner.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * @alpha
+ */
+export interface QueryRunner {
+  getData(options: QueryRunnerGetDataOptions): Observable<PanelData>;
+  run(options: QueryRunnerOptions): Promise<void>;
+  cancelQuery(): void;
+  resendLastResult(): void;
+  destroy(): void;
+  useLastResultFrom(runner: QueryRunner): void;
+  getLastResult(): PanelData | undefined;
```
Comment (torkelo) [inline packages/grafana-data/src/types/queryRunner.ts:57]: think we can remove resendLastResult, useLastResultFrom and getLastResult (it make them internal to core app, only on the implementation). 

I think we can move this interface to runtime.

File: packages/grafana-data/src/types/queryRunner.ts (hunk 2 above)
Comment (mckn) [reply packages/grafana-data/src/types/queryRunner.ts:14]: > hmm, we might be able to change the interface but let's do that in a different PR. I would like to keep this as minimal as possible.
> 
> So the benefit of having this in `@grafana/data` is that we can have a dependency on this interface in e.g. `@grafana/ui` which might be beneficial. We had this problem with the `DataSourceSrv`. If that interface would have been in `@grafana/data` we could have moved that component to `@grafana/ui`.

@torkelo regarding keeping it in `@grafana/data`. Wdyt?

File: packages/grafana-data/src/types/queryRunner.ts (hunk 5 above)
Comment (mckn) [reply packages/grafana-data/src/types/queryRunner.ts:57]: OK ok, will see if I can remove them.

**Comment by mckn** [comment]
Closing this in favor of #31901

**Pull request closed without merging**